### PR TITLE
Implement coroutine generator support

### PR DESCRIPTION
**Pull request by CaseyCarter**
* Cmake support for detecting `-fcoroutines-ts` compiler flag and library support in `<experimental/coroutine>`
* Config system support for `RANGES_CXX_COROUTINES`
* Implement coroutine Range generator `ranges::experimental::generator`, and SizedRange generator `ranges::experimental::sized_generator`

### CODE REVIEW

**Comment by ericniebler** [comment]
Wow! I need to work through the generator code to understand what it's doing. I would also be willing to accept patches to make ranges::optional be more like std::optional.

File: include/range/v3/experimental/utility/generator.hpp [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+            } // namespace coroutine_owner_
+        } // namespace detail
+        /// \endcond
+
+        namespace experimental
+        {
+            // An owning coroutine_handle
+            template<typename Promise>
+            struct coroutine_owner
+            : std::experimental::coroutine_handle<Promise>
```
Comment (CaseyCarter) [inline include/range/v3/experimental/utility/generator.hpp]: Dammit - VS Code broke all of these partial indent lines when I moved this stuff in an indent for the `experimental` namespace, and I missed it.

**Comment by CaseyCarter** [comment]
Food for thought: I implemented `coro_fn` which wraps any viewable input range into a coroutine for testing purposes, but didn't provide that tool to users. Does anyone see a use case for such a tool?

**Comment by Manu343726** [comment]
Do you know of any library implementing this generator with boost coroutine or coroutine2? Would you be interested in such a generator backend for this library?

**Comment by CaseyCarter** [comment]
> Do you know of any library implementing this generator with boost coroutine or coroutine2? Would you be interested in such a generator backend for this library?

I don't know of a library implementing this generator atop any flavor of boost coroutine, nor have we had demand for such a thing. (I don't think people are aware that `ranges::experimental::generator` exists.) That said, if there's someone out there that would like to use generators - you really can't beat the ease of range definition - who doesn't have access to TS coroutines, I imagine such an backend would be useful to them.

**Comment by Manu343726** [comment]
I'm asking because I would like to use ranges to filter visitation results, where the visitor is the usual tree recursive visitor:

``` cpp
visit_tree(root, [](const node& node)
{
    ...
});

// with coroutines:

void visit(push_t& yield, const node& root)
{
    visit(root, [&yield](const node& node)
    {
        yield(node);
    });
}

pull_t coro{[&root](push_t& yield)
{
    visit(yield, root);  
}};

auto nodes_named_a = coro | filter([](const node& node)
{
    return node.name == "a";
});
```

In case of boost coroutines, since the pull type implements begin() and end() maybe I could pass a coroutine directly to ranges (Like in the example above) and it would work. I'm not sure if that works directly.

**Pull request merged**